### PR TITLE
feat: Allow local certificates to be used by Caddy

### PR DESCRIPTION
**Pull request by thePanz**
Relates to #79  and #93

### CODE REVIEW

**Comment by B-Galati** [comment]
Great! Thank you @thePanz :pray:

**Comment by dunglas** [comment]
Awesome! As it's a relatively specific use case, I would prefer to not add too much "unnecessary" comments to the skeleton itself. Could we just add a doc entry explaining what lines to add as we've done for XDebug? We can use the ` ```patch ` Markdown trick for this purpose.

**Comment by maxhelias** [comment]
Wow! Thanks @thePanz  🎉 

@dunglas, maybe rework the xdebug part to have only one page on development options (xdebug + tls)

**Comment by dunglas** [comment]
@maxhelias, I'm not sure. I think having small pages with descriptive URLs and titles will improve SEO.

**Comment by zmitic** [comment]
After many hours, I failed to make it work. My setup is Linux mint, and all I get is non-responsive page.

Troubleshooting page with ``docker cp`` command also doesn't work, even with changed path (usr/share instead of usr/local/share).

I am out of ideas.

**Comment by mfehr94** [comment]
@zmitic the problem is the `caddy:80` in `SERVER_NAME="api.localhost, caddy:80"`. It is not possible to configure tls on sites that listen on the HTTP port.

@thePanz @maxhelias @dunglas any suggestions to solve this issue?

**Comment by Jayfrown** [comment]
Thanks, I got it to work as follows:

```diff
❯ git diff --staged | cat        
diff --git a/docker-compose.override.yml b/docker-compose.override.yml
index 0143fec..2ab2b3d 100644
--- a/docker-compose.override.yml
+++ b/docker-compose.override.yml
@@ -17,6 +17,10 @@ services:
     volumes:
       - ./docker/caddy/Caddyfile:/etc/caddy/Caddyfile:ro
       - ./public:/srv/app/public:ro
+      - ./docker/caddy/certs:/etc/caddy/certs:ro
+    environment:
+      SERVER_NAME: ${SERVER_NAME:-localhost, caddy}
+      CADDY_TLS_CONFIG: "tls /etc/caddy/certs/tls.pem /etc/caddy/certs/tls.key"
 
   phpldapadmin:
     image: osixia/phpldapadmin:latest
diff --git a/docker-compose.yml b/docker-compose.yml
index c1a1ea9..044033f 100644
--- a/docker-compose.yml
+++ b/docker-compose.yml
@@ -20,7 +20,7 @@ services:
       start_period: 30s
     environment:
       DATABASE_URL: "mysqli://root:${MYSQL_ROOT_PASSWORD:-!ChangeMe!}@database:3306/${MYSQL_DATABASE:-app}?serverVersion=mariadb-10.3.13"
-      MERCURE_URL: ${CADDY_MERCURE_URL:-http://caddy/.well-known/mercure}
+      MERCURE_URL: ${CADDY_MERCURE_URL:-https://caddy/.well-known/mercure}
       MERCURE_PUBLIC_URL: https://${SERVER_NAME:-localhost}/.well-known/mercure
       MERCURE_JWT_SECRET: ${CADDY_MERCURE_JWT_SECRET:-!ChangeMe!}
       JWT_PASSPHRASE: ${JWT_PASSPHRASE:-!ChangeMe!}
diff --git a/docker/caddy/Caddyfile b/docker/caddy/Caddyfile
index 6d5387c..18f70c3 100644
--- a/docker/caddy/Caddyfile
+++ b/docker/caddy/Caddyfile
@@ -11,6 +11,8 @@
 
 {$SERVER_NAME}
 
+{$CADDY_TLS_CONFIG}
+
 log
 
 route {
diff --git a/docker/caddy/certs/.gitignore b/docker/caddy/certs/.gitignore
new file mode 100644
index 0000000..65e3d29
--- /dev/null
+++ b/docker/caddy/certs/.gitignore
@@ -0,0 +1,3 @@
+# Ignore locally generated certs for Caddy
+tls.key
+tls.pem
```

I added the `.gitignore` because in a multi-developer project you would want everyone to have their own locally generated CA and certs.

---

Now caddy uses the locally generated certificates, which means the cert is trusted within my browser. However, the CA generated by `mkcert` is apparently not automatically trusted in my local PHP set-up, so local commands that try to send Mercure updates still fail as follows:

```
SSL certificate problem: unable to get local issuer certificate for "https://localhost/.well-known/mercure".
```

As the CA is now (relatively) static I can add it to the trust store used by PHP as follows:

```
cat "$(mkcert -CAROOT)/rootCA.pem" >> /usr/local/etc/openssl@1.1/cert.pem
```

Note that this might be a MacOS / brew PHP specific issue. Can someone confirm/refute whether the CA generated by `mkcert` is automatically trusted in local PHP when run on Linux?

**Comment by Jayfrown** [comment]
Both issues could be fixed if there is a separate configuration within the Caddyfile for caddy:80 and localhost:443, such that both local PHP and the docker PHP app can send updates to `http://caddy/.well-known/mercure` (as is the default for the dockerized PHP app). The stanza for localhost:443 could then use the certs generated by `mkcert`.

**Comment by mario-fehr** [comment]
Any plans to fix and merger this?

**Comment by maxhelias** [comment]
I updated this PR

**Comment by thePanz** [comment]
Thanks @maxhelias for the update, sorry for not following up on this :+1:

**Comment by dunglas** [comment]
This looks good to me.
@mholt would you mind checking if it's in sync with Caddy's best practices?

File: docs/tls.md [hunk 1]
```diff
@@ -0,0 +1,38 @@
+# TLS Certificates
+
+## Using caddy TLS certificates
+
+With a basic installation, the self-signed certificate from caddy is not know by your local machine. 
+You must copy it and add it to the trusted certificates :
```
Comment (dunglas) [inline docs/tls.md:6]:
```suggestion
You must add the authority to the trust store of the host:
```

File: docker/caddy/Caddyfile [hunk 2]
```diff
@@ -5,6 +5,8 @@
 
 {$SERVER_NAME}
 
+{$CADDY_TLS_CONFIG}
```
Comment (dunglas) [inline docker/caddy/Caddyfile:8]:
```suggestion
{$CADDY_EXTRA_CONFIG}
```

As this may have other uses.

File: docs/tls.md [hunk 3]
```diff
@@ -0,0 +1,38 @@
+# TLS Certificates
+
+## Using caddy TLS certificates
+
+With a basic installation, the self-signed certificate from caddy is not know by your local machine.
```
Comment (dunglas) [inline docs/tls.md:5]:
```suggestion
With a standard installation, the authority used to sign certificates generated in the Caddy container is not trusted by your local machine. 
```

File: docs/tls.md [hunk 4]
```diff
@@ -0,0 +1,38 @@
+# TLS Certificates
+
+## Using caddy TLS certificates
```
Comment (dunglas) [inline docs/tls.md:3]:
```suggestion
## Trusting the Authority
```

File: docs/tls.md [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+```
+# Mac
+$ docker cp $(docker compose ps -q caddy):/data/caddy/pki/authorities/local/root.crt /tmp/root.crt && sudo security add-trusted-cert -d -r trustRoot -k /Library/Keychains/System.keychain /tmp/root.crt
+# Linux
+$ docker cp $(docker compose ps -q caddy):/data/caddy/pki/authorities/local/root.crt /usr/local/share/ca-certificates/root.crt && sudo update-ca-certificates
+# Windows
+$ docker compose cp caddy:/data/caddy/pki/authorities/local/root.crt %TEMP%/root.crt && certutil -addstore -f "ROOT" %TEMP%/root.crt
+```
+
+## Using local TLS certificates
```
Comment (dunglas) [inline docs/tls.md:17]:
```suggestion
## Using Custom TLS Certificates
```

File: docs/tls.md [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+Caddy is configured to generate TLS certificates by using Let's Encrypt, but for local deployments this does not work,
```
Comment (dunglas) [inline docs/tls.md:19]:
```suggestion
By default, Caddy will automatically generate TLS certificates using Let's Encrypt or ZeroSSL.
```

File: docs/tls.md [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+and the internal caddy certificate should be imported locally (see above).
```
Comment (dunglas) [inline docs/tls.md:20]:
```suggestion
But sometimes you may prefer using custom certificates.
```

File: docs/tls.md [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+
+To use local certificates created with [mkcert](https://github.com/FiloSottile/mkcert) do as follows:
```
Comment (dunglas) [inline docs/tls.md:22]:
```suggestion
For instance, to use self-signed certificates created with [mkcert](https://github.com/FiloSottile/mkcert) do as follows:
```

File: docs/tls.md [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+1. Locally install `mkcert`
+2. Create the folder storing the certs: 
+   `mkdir docker/caddy/certs -p`
+3. Generate the certificates for your local host (example: "server-name.localhost"):
+   `mkcert -cert-file docker/caddy/certs/tls.pem -key-file docker/caddy/certs/tls.key "server-name.localhost"`
+4. Add these lines to the `./docker-compose.override.yml` file about `CADDY_TLS_CONFIG` environment and volume for the `caddy` service :
+    ```diff
+    caddy:
+    +  environment:
+    +    CADDY_TLS_CONFIG: "tls /etc/caddy/certs/tls.pem /etc/caddy/certs/tls.key"
```
Comment (dunglas) [inline docs/tls.md:33]:
```suggestion
    +    CADDY_EXTRA_CONFIG: "tls /etc/caddy/certs/tls.pem /etc/caddy/certs/tls.key"
```

File: docs/troubleshooting.md [hunk 10]
```diff
@@ -4,18 +4,11 @@
...
-    $ docker cp $(docker compose ps -q caddy):/data/caddy/pki/authorities/local/root.crt /usr/local/share/ca-certificates/root.crt && sudo update-ca-certificates
-    # Windows
-    $ docker compose cp caddy:/data/caddy/pki/authorities/local/root.crt %TEMP%/root.crt && certutil -addstore -f "ROOT" %TEMP%/root.crt
-
 ## HTTPs and Redirects
 
 If Symfony is generating an internal redirect for an `https://` url, but the resulting url is `http://`, you have to uncomment the `TRUSTED_PROXIES` setting in your `.env` file.
 For more details see the [Symfony internal redirect documentation](https://symfony.com/doc/current/routing.html#redirecting-to-urls-and-routes-directly-from-a-route).
+
+## TLS trust issues
```
Comment (dunglas) [inline docs/troubleshooting.md:12]:
```suggestion
## TLS/HTTPS Issues
```

**Review by dunglas — CHANGES_REQUESTED** [review]

**Review by mholt — APPROVED** [review]
I think that looks good overall!

The only thing is that we pretty much always recommend letting Caddy manage the certificates, instead of using external tools or certs, whenever possible. (It should almost always be possible to use Caddy's certs these days, even if you need a custom CA or something.)

**Pull request merged**